tor/controller.go

```diff
@@ -176,6 +176,13 @@ func (c *Controller) authenticate() error {
 			"%v", err)
 	}
 
+	// If cookie is empty and there's no error, we have a NULL
+	// authentication method that we should use instead.
+	if len(cookie) == 0 {
```

Comment:
As we introduce more auth methods, perhaps we should consider refactoring extracting the supported auth methods from this method into another one? So the flow would be something like:

```go
func authenticate() {
    supportedAuthMethods := getSupportedAuthMethods()
    switch {
    case len(cookie) == 0 && supportedAuthMethods.Contains("NULL"):
        // Proceed with NULL auth.
    case len(cookie) != 0 && supportedAuthMethods.Contains("SAFECOOKIE"):
        // Proceed with cookie auth.
    }
}
```